[SPARK-32850][CORE][K8S] Simplify the RPC message flow of decommission

Ngone51 opened the pull request:
<!--
Thanks for sending a pull request!  Here are some tips for you:
  1. If this is your first time, please read our contributor guidelines: https://spark.apache.org/contributing.html
  2. Ensure you have added or run the appropriate tests for your PR: https://spark.apache.org/developer-tools.html
  3. If the PR is unfinished, add '[WIP]' in your PR title, e.g., '[WIP][SPARK-XXXX] Your PR title ...'.
  4. Be sure to keep the PR description updated to reflect all changes.
  5. Please write your PR title to summarize what this PR proposes.
  6. If possible, provide a concise example to reproduce the issue for a faster review.
  7. If you want to add a new configuration, please read the guideline first for naming configurations in
     'core/src/main/scala/org/apache/spark/internal/config/ConfigEntry.scala'.
-->

### What changes were proposed in this pull request?
<!--
Please clarify what changes you are proposing. The purpose of this section is to outline the changes and how this PR fixes the issue. 
If possible, please consider writing useful notes for better and faster reviews in your PR. See the examples below.
  1. If you refactor some codes with changing classes, showing the class hierarchy will help reviewers.
  2. If you fix some SQL features, you can provide some references of other DBMSes.
  3. If there is design documentation, please add the link.
  4. If there is a discussion in the mailing list, please add the link.
-->


This PR cleans up the RPC message flow among the multiple decommission use cases, it includes changes:

* Keep `Worker`'s decommission status be consistent between the case where decommission starts from `Worker` and the case where decommission starts from the `MasterWebUI`: sending `DecommissionWorker` from `Master` to `Worker` in the latter case.

* Change from two-way communication to one-way communication when notifying decommission between driver and executor: it's obviously unnecessary for the executor to acknowledge the decommission status to the driver since the decommission request is from the driver. And it's same in reverse.

* Only send one message instead of two(`DecommissionSelf`/`DecommissionBlockManager`) when decommission the executor: executor and `BlockManager` are in the same JVM.

* Clean up codes around here.



### Why are the changes needed?
<!--
Please clarify why the changes are needed. For instance,
  1. If you propose a new API, clarify the use case for a new API.
  2. If you fix a bug, you can clarify why it is a bug.
-->

Before:

<img width="1948" alt="WeChat56c00cc34d9785a67a544dca036d49da" src="https://user-images.githubusercontent.com/16397174/92850308-dc461c80-f41e-11ea-8ac0-287825f4e0c4.png">

After:
<img width="1968" alt="WeChat05f7afb017e3f0132394c5e54245e49e" src="https://user-images.githubusercontent.com/16397174/93189571-de88dd80-f774-11ea-9300-1943920aa27d.png">




(Note the diagrams only counts those RPC calls that needed to go through the network. Local RPC calls are not counted here.)

After this change, We reduced 6 original RPC calls and added one more RPC call for keeping the consistent decommission status for the Worker. And the RPC flow becomes more clear.


### Does this PR introduce _any_ user-facing change?
<!--
Note that it means *any* user-facing change including all aspects such as the documentation fix.
If yes, please clarify the previous behavior and the change this PR proposes - provide the console output, description and/or an example to show the behavior difference if possible.
If possible, please also clarify if this is a user-facing change compared to the released Spark versions or within the unreleased branches such as master.
If no, write 'No'.
-->

No.


### How was this patch tested?
<!--
If tests were added, say they were added here. Please make sure to add some test cases that check the changes thoroughly including negative and positive cases if possible.
If it was tested in a way different from regular unit tests, please clarify how you tested step by step, ideally copy and paste-able, so that other reviewers can test and check, and descendants can verify in the future.
If tests were not added, please describe why they were not added and/or why it was difficult to add.
-->

Updated existing tests.

Ngone51 committed:
Revert "Revert "[SPARK-32850][CORE] Simplify the RPC message flow of decommission""

This reverts commit 0c66813ad9867e366689b47c81bdd8a94ac17828.

Ngone51 commented:
Bring this back since it isn't the original commit that breaks the K8S test. As the PR https://github.com/apache/spark/pull/29751 merged before this already failed the K8S tests

SparkQA commented:
**[Test build #128934 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/128934/testReport)** for PR 29817 at commit [`5ca0fe8`](https://github.com/apache/spark/commit/5ca0fe88f5b651d4ead5abf157ec70cbd77fac13).

HyukjinKwon commented:
@holdenk, seems like the test failure wasn't caused by this PR. Dose your -1 at [here](https://issues.apache.org/jira/browse/SPARK-32937?focusedCommentId=17198867&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-17198867) still stand?

cc @dongjoon-hyun as well per #29408.

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/33556/

holdenk commented:
The -1 is only on the PR that I mentioned it on (the refactoring with many sub classes) which still stands. I would like to review this PR more though since I think we probably need better test coverage of this change than it originally had. Sound good? Thanks for checking in about that :)

SparkQA commented:
**[Test build #128935 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/128935/testReport)** for PR 29817 at commit [`15f6085`](https://github.com/apache/spark/commit/15f6085d24bd26fc803831a91a424c94e3bf10f6).

HyukjinKwon commented:
Got it about -1 but what about we pushing this as is, and work on the test as followups? It's a bit odds that we reverted it for the reason this PR didn't cause, and ask more things before merging it in.

SparkQA commented:
Kubernetes integration test status failure
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/33556/

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/33556/
Test FAILed.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/33557/
Test FAILed.

holdenk commented:
I think we should not commit this with the K8s test being broken. It’s in the same chunk of code and changes the logging string (although another PR also changed that string first too?). I do not believe this PR was appropriately tested when first merger given it changed decommissioning messages and did not run the decommission tests.

For clarity: if you want to fix the tests in a separate PR that’s ok with me, but I would prefer not to commit this without passing integration testing.

SparkQA commented:
**[Test build #128934 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/128934/testReport)** for PR 29817 at commit [`5ca0fe8`](https://github.com/apache/spark/commit/5ca0fe88f5b651d4ead5abf157ec70cbd77fac13).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/128934/
Test PASSed.

SparkQA commented:
**[Test build #128935 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/128935/testReport)** for PR 29817 at commit [`15f6085`](https://github.com/apache/spark/commit/15f6085d24bd26fc803831a91a424c94e3bf10f6).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `  case class DecommissionWorkers(ids: Seq[String]) extends DeployMessage`
  * `  case class WorkerDecommissioning(id: String, workerRef: RpcEndpointRef) extends DeployMessage`
  * `  case class ExecutorDecommissioning(executorId: String) extends CoarseGrainedClusterMessage`

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/128935/
Test PASSed.

Ngone51 commented:
> I think we should not commit this with the K8s test being broken.

We don't. That's also why I added `[K8S]` tag in the PR title. And feel free to leave comments, I can address them in followups.

HyukjinKwon commented:
@holdenk, why don't you take a look for a test failure since it blocks all changes in decommission in k8s, and you were involved mainly in the development there?

cloud-fan commented:
Can we find out which commit caused the test failure in the first place? We should either revert that commit, or fix it soon, as the test failure blocks others.

Since this PR is resubmitted (although the revert is not necessary now given the test failure was already there), I think it's a good chance for @holdenk to take a closer look before re-merging. And I agree with @holdenk that we can't merge a PR when the related test is already broken. We should fix that first. @holdenk can you give some hints about it? I took a look at https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/33556/ , but I don't even see how the test failed. The output is very different from normal Spark tests.

holdenk commented on core/src/main/scala/org/apache/spark/deploy/worker/Worker.scala:
```diff
@@ -70,7 +70,10 @@ private[deploy] class Worker(
     SignalUtils.register("PWR", "Failed to register SIGPWR handler - " +
-      "disabling worker decommission feature.")(decommissionSelf)
+      "disabling worker decommission feature.") {
+       self.send(WorkerSigPWRReceived)
```
I think this might mean we return from handling the signal right away rather than waiting for decommissionSelf to be finished. Is this an intentional change?

Also this will no longer report a decommissioning failure by signal return value, so may block pod deletion or other cleanup task longer than needed.

Ngone51 replied:
This's Worker, I guess you cares more about the executor? In Worker, `decommissionSelf` always returns true. and In exeutor, there's a change to return false to fail the decommissionSelf but seems rarely happen. If you would insist on returning the value, I think we can use askSync instead.

holdenk replied:
Can you look into what the difference of this behavior might cause at the system level and then tell me if that’s a desired change? I’m ok with us making changes here, I just want us to be intentional and know if we need to test the change and it seems like this change was incidental.

Ngone51 replied:
The return value of the signal handling decides whether we should forward the signal to the other handlers. If `true`, no other handlers will handle the `PWR` signal except ourselves. If `false`, we will handle it (for decommission) and other handlers will handle it too.  Do you expect other handlers to continue handling the `SIGPWR` when the system isn't really experiencing a power failure?

holdenk replied:
I do, I think if the signal is unhandled then the process will be killed immediately. If we think of decommissioning/graceful shut down I believe that behavior is desirable, since if we can't shut down gracefully the least we can do is exit quickly.

Ngone51 replied:
Updated, but please note that I only updated for the executor's case since Worker's case always returns `true` before.

holdenk commented on core/src/main/scala/org/apache/spark/deploy/worker/Worker.scala:
```diff
@@ -668,8 +672,13 @@ private[deploy] class Worker(
+
+    case WorkerSigPWRReceived =>
       decommissionSelf()
+      // Tell master we starts decommissioning so it stops trying to launch executor/driver on us
```
May I suggest "Tell the master that we are starting decommissioning so it stops trying to launch executor/driver on us"

holdenk commented on core/src/main/scala/org/apache/spark/executor/CoarseGrainedExecutorBackend.scala:
```diff
@@ -79,12 +79,17 @@ private[spark] class CoarseGrainedExecutorBackend(
-    SignalUtils.register("PWR", "Failed to register SIGPWR handler - " +
-      "disabling decommission feature.")(decommissionSelf)
+    if (env.conf.get(DECOMMISSION_ENABLED)) {
+      logInfo("Registering PWR handler to trigger decommissioning.")
+      SignalUtils.register("PWR", "Failed to register SIGPWR handler - " +
+      "disabling executor decommission feature.") {
+        self.send(ExecutorSigPWRReceived)
+        true
+      }
+    }
```
Similar comments as in Worker.scala

holdenk commented on core/src/main/scala/org/apache/spark/executor/CoarseGrainedExecutorBackend.scala:
```diff
@@ -166,17 +171,6 @@ private[spark] class CoarseGrainedExecutorBackend(
       if (executor == null) {
         exitExecutor(1, "Received LaunchTask command but executor was null")
       } else {
-        if (decommissioned) {
-          val msg = "Asked to launch a task while decommissioned."
-          logError(msg)
-          driver match {
-            case Some(endpoint) =>
-              logInfo("Sending DecommissionExecutor to driver.")
-              endpoint.send(DecommissionExecutor(executorId, ExecutorDecommissionInfo(msg)))
-            case _ =>
-              logError("No registered driver to send Decommission to.")
-          }
-        }
```
I don't think we should just take this out. async sends could fail, re-sending the message if we receive a request which indicates the master hasn't received our notification indicates we should resend.

Ngone51 replied:
First, we use `askSync` to send decommission notice to the driver whenever it needs(see `ExecutorSigPWRReceived`). Second, even if driver receives the decommission notice successfully, there still could be LaunchTask request due to the async between LaunchTask and decommission notice. Third, this part also uses async send,  so we still can not ensure the decommission notice is received by driver successfully.

holdenk replied:
Right, so we should resend the notice then right?

Ngone51 replied:
No. Sorry if I didn't explain it clearly.

We should already send decommission notice to the driver when `decommissioned = true`, which using `askSync`. If `askSync` still fail, I wouldn't expect another send would succeed.

cloud-fan replied:
Yea I don't see the point of resending the notice to the driver, especially in this race condition. If we want to make sure the driver is noticed, we should design a mechanism for it, instead of doing it here randomly.

holdenk commented on core/src/main/scala/org/apache/spark/executor/CoarseGrainedExecutorBackend.scala:
```diff
@@ -213,9 +207,17 @@ private[spark] class CoarseGrainedExecutorBackend(
+    case ExecutorSigPWRReceived =>
+      decommissionSelf()
+      if (driver.nonEmpty) {
+        // Tell driver we starts decommissioning so it stops trying to schedule us
```
s/we starts/we are starting/
?

holdenk commented on core/src/main/scala/org/apache/spark/executor/CoarseGrainedExecutorBackend.scala:
```diff
@@ -213,9 +207,17 @@ private[spark] class CoarseGrainedExecutorBackend(
+
+    case ExecutorSigPWRReceived =>
+      decommissionSelf()
+      if (driver.nonEmpty) {
```
Doing this after decommissionSelf introduces a race condition, not too important one but given that it wasn't in the original flow I'd suggest swapping the order.

Ngone51 replied:
What's race condition?

holdenk replied:
So we don’t ask the driver to stop scheduling jobs on us first, and the driver could ask us to run a job while we are part way through decommissioning. This won’t result in a failure because well accept the job but it will slow down the decommissioning. So swap the order of these two.

Ngone51 replied:
Addressed

holdenk commented on core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedSchedulerBackend.scala:
```diff
@@ -272,10 +268,14 @@ class CoarseGrainedSchedulerBackend(scheduler: TaskSchedulerImpl, val rpcEnv: Rp
-        context.reply(decommissionExecutor(executorId, decommissionInfo,
-          adjustTargetNumExecutors = false))
+      case ExecutorDecommissioning(executorId) =>
+        logWarning(s"Received executor $executorId decommissioned message")
```
Here might be where you break the test suite last time, so double check it.

holdenk commented on core/src/main/scala/org/apache/spark/storage/BlockManager.scala:
```diff
@@ -1809,7 +1809,9 @@ private[spark] class BlockManager(
   }
 
-  def decommissionBlockManager(): Unit = synchronized {
+  def decommissionBlockManager(): Unit = storageEndpoint.ask(DecommissionBlockManager)
```
Why?

Ngone51 replied:
?

holdenk replied:
Why did you make this change?

Ngone51 replied:
If I unsterstand your question correctly:

We didn't really change the `decommissionBlockManager`. The original `decommissionBlockManager` has been renamed to `decommissionSelf` to avoid the naming collision.

holdenk replied:
Makes sense, although maybe introducing a new name instead of changing the use of a previous function name would be easier for verifying.

holdenk reviewed (commented):
Thanks for reverting this and resubmitting it. I know you believe the original PR didn't cause the test failure, but that's only half true, this PR just broke the test some more.

That being said I still have concerns this PR is not sufficiently tested, can you add some more tests for the new flows you've introduced?

holdenk commented:
I really like the idea of simplifying the RPC message flow, thanks for taking this on @Ngone51 and I'm sorry the code here is so brittle to these types of changes (the K8s integration tests are kind of limited).

HyukjinKwon commented:
> but that's only half true, this PR just broke the test some more.

@holdenk, you're kidding right? There was only one test failure that was not caused by this PR in K8S tests. That test was just fixed by you. How come this PR broke more tests? Can you be more explicit on that? Which tests were more broken, and how did you test?

holdenk commented:
I’m not joking. The test has multiple conditions. Another PR broke one of the conditions. This PR in its original broke another one of the conditions. It was reported in the same test because K8s integration tests focus on the integration so they cover multiple pieces. If you want I’m happy to do a video call (and I can be flex on my timezone if that’s a constraint) and screen share so we can discuss the details.

holdenk commented:
If it helps I called out the part of the PR I believe most likely responsible for that during my code review already.

Ngone51 commented:
> This PR in its original broke another one of the conditions.

Except for the breaking of the log(like you just fixed), what other conditions this PR breaks?

Ngone51 commented on core/src/main/scala/org/apache/spark/deploy/DeployMessage.scala:
```diff
@@ -61,13 +61,34 @@ private[deploy] object DeployMessages {
+  object DecommissionWorker extends DeployMessage
+
+  /**
+   * A message that sent to the Worker itself when it receives PWR signal,
```
yeah

Ngone51 commented:
> That being said I still have concerns this PR is not sufficiently tested, can you add some more tests for the new flows you've introduced?

There's only one new flow that is from Master to Worker. I can update the existing test by verifying Worker's decommission status... What kind of other concerns do you have? Could you elaborate more? So I can improve the PR accordingly.

holdenk commented:
I’m mostly concerned with the change around how the storage decommissioning is being done now, I’d like to see some tests that the flow from the master to the worker results in storage decommissioning.

holdenk commented:
Also there are multiple log conditions, this PR broke one of them. Another PR had broken another one.

HyukjinKwon commented:
@Ngone51, shall we add a test https://github.com/apache/spark/pull/29817#issuecomment-698438316 and fix the test as requested? Seems like otherwise good to go.

Ngone51 committed:
update comment

Ngone51 committed:
update comment

Ngone51 committed:
switch order

Ngone51 committed:
update test

Ngone51 committed:
add test to verify BlockManager decommission flow

Ngone51 committed:
update log

SparkQA commented:
**[Test build #129172 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/129172/testReport)** for PR 29817 at commit [`23bfdf5`](https://github.com/apache/spark/commit/23bfdf5a5bcc05b4c8066593aa1607dbe7dbe5b0).

Ngone51 commented:
@holdenk Thanks for the review! I've addressed your most comments. And tests are updated and added in fa04b493324ea52698bed7ce15795530052b37be and
9d0f36d73e5750267ec2f33e7d964fc09778767f .

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/33787/

SparkQA commented:
Kubernetes integration test status failure
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/33787/

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/33787/
Test FAILed.

SparkQA commented:
**[Test build #129172 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/129172/testReport)** for PR 29817 at commit [`23bfdf5`](https://github.com/apache/spark/commit/23bfdf5a5bcc05b4c8066593aa1607dbe7dbe5b0).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/129172/
Test PASSed.

Ngone51 committed:
fix

SparkQA commented:
**[Test build #129182 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/129182/testReport)** for PR 29817 at commit [`e3ca2af`](https://github.com/apache/spark/commit/e3ca2aff082ecaf2850030cab56fe083d28142f2).

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/33795/

SparkQA commented:
Kubernetes integration test status failure
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/33795/

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/33795/
Test FAILed.

SparkQA commented:
**[Test build #129182 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/129182/testReport)** for PR 29817 at commit [`e3ca2af`](https://github.com/apache/spark/commit/e3ca2aff082ecaf2850030cab56fe083d28142f2).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/129182/
Test FAILed.

HyukjinKwon commented:
retest this please

SparkQA commented:
**[Test build #129198 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/129198/testReport)** for PR 29817 at commit [`e3ca2af`](https://github.com/apache/spark/commit/e3ca2aff082ecaf2850030cab56fe083d28142f2).

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/33813/

SparkQA commented:
Kubernetes integration test status failure
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/33813/

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/33813/
Test FAILed.

holdenk commented:
Tests are still failing @Ngone51 happy to take another look once the K8s integration tests are passing.

SparkQA commented:
**[Test build #129198 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/129198/testReport)** for PR 29817 at commit [`e3ca2af`](https://github.com/apache/spark/commit/e3ca2aff082ecaf2850030cab56fe083d28142f2).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/129198/
Test PASSed.

Ngone51 committed:
fix

Ngone51 committed:
address comment

Ngone51 committed:
update

SparkQA commented:
**[Test build #129277 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/129277/testReport)** for PR 29817 at commit [`af9dbdd`](https://github.com/apache/spark/commit/af9dbdd4dc96e5e3f2ed2e9e21bb803081337df9).

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/33894/

SparkQA commented:
Kubernetes integration test status failure
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/33894/

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/33894/
Test FAILed.

SparkQA commented:
**[Test build #129277 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/129277/testReport)** for PR 29817 at commit [`af9dbdd`](https://github.com/apache/spark/commit/af9dbdd4dc96e5e3f2ed2e9e21bb803081337df9).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/129277/
Test PASSed.

holdenk commented on core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedSchedulerBackend.scala:
```diff
@@ -465,72 +464,50 @@ class CoarseGrainedSchedulerBackend(scheduler: TaskSchedulerImpl, val rpcEnv: Rp
-      adjustTargetNumExecutors: Boolean): Seq[String] = {
-
+      adjustTargetNumExecutors: Boolean,
+      triggeredByExecutor: Boolean): Seq[String] = withLock {
```
Not 100% sure, but I do remember that the locking mechanism withLock did manage to deadlock before, it might be good to add a logging statement in here that verifies this block is called in the integration test.

Ngone51 replied:
> Not 100% sure, but I do remember that the locking mechanism withLock did manage to deadlock before

Yeah. We need to avoid the deadlock now because we need both CoarseGrainedSchedulerBackend's lock and TaskSchedulerImpl's lock within the same code block. 


> it might be good to add a logging statement in here that verifies this block is called in the integration test.

Updated the logs and the integration test.

holdenk commented on core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedSchedulerBackend.scala:
```diff
@@ -465,72 +464,50 @@ class CoarseGrainedSchedulerBackend(scheduler: TaskSchedulerImpl, val rpcEnv: Rp
-          logError("Unexpected error during block manager " +
-            s"decommissioning for executor $executorId: ${e.toString}", e)
-          return false
+    if (!triggeredByExecutor) {
```
I think we might want to move this up because doing the storage decommissioning before the executor it's self is marked for decommissioning could mean that local blocks get stored in a manner where we might not know to migrate them. If you've dug through the code here and you don't believe that happens that's fine.

Ngone51 replied:
We don't really migrate blocks when decommissioning BlockManagers above. We only mark them as being decommissioning at driver side. So I think the problem you mentioned won't exist.

holdenk commented on core/src/main/scala/org/apache/spark/storage/BlockManagerMasterEndpoint.scala:
```diff
@@ -163,8 +163,7 @@ class BlockManagerMasterEndpoint(
 
     case DecommissionBlockManagers(executorIds) =>
-      val bmIds = executorIds.flatMap(blockManagerIdByExecutor.get)
-      decommissionBlockManagers(bmIds)
+      decommissioningBlockManagerSet ++= executorIds.flatMap(blockManagerIdByExecutor.get)
```
Maybe we can add a comment similar to the previous one of decommissionBlockManagers stating why we are doing this and it's intended effect.

Ngone51 replied:
Added.

holdenk commented on core/src/main/scala/org/apache/spark/deploy/worker/Worker.scala:
```diff
@@ -668,8 +672,14 @@ private[deploy] class Worker(
       decommissionSelf()
+      // Tell the Master that we are starting decommissioning
+      // so it stops trying to launch executor/driver on us
+      sendToMaster(WorkerDecommissioning(workerId, self))
```
To make this code testable I'd move it down to inside of decommissionSelf and give decommissionSelf a param for if the decom originated locally.

Ngone51 replied:
I added the param previously but removed it later in order to keep `decommissionSelf` unchanged according to the comment: https://github.com/apache/spark/pull/29722#discussion_r487855223

holdenk commented on core/src/test/scala/org/apache/spark/deploy/client/AppClientSuite.scala:
```diff
@@ -122,7 +122,10 @@ class AppClientSuite
-      workers.foreach(worker => worker.decommissionSelf())
+      workers.foreach { worker =>
+       worker.decommissionSelf()
+       master.self.send(WorkerDecommissioning(worker.workerId, worker.self))
```
left a comment up above about how to improve the testing, I think that part of the test should cover the message being sent from the worker so we probably don't want to manually send that message here.

holdenk commented on resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/DecommissionSuite.scala:
```diff
@@ -41,9 +41,8 @@ private[spark] trait DecommissionSuite { k8sSuite: KubernetesSuite =>
-        "Received decommission executor message",
-        "Acknowledged decommissioning block manager",
-        ": Executor decommission.",
+        "Asking BlockManagers", "to decommissioning",
+        "Asking executor", "to decommissioning.",
```
nit: "to decommissioning." and "to decommissioning." are close enough you really only need the second to cover both cases since we're looking for the string anywhere. If you did want to you could switch this to regex and do more accurate matching. up to you.

Ngone51 replied:
Updated to use only one "decommissioning".

holdenk reviewed (commented):
Just a few notes, I noticed the integration tests are still failing. Let me know if you are stuck and would like some help figuring out what's going with the integration tests.

Ngone51 commented:
@holdenk Sorry for the delay. I had taken some days off recently. I'm addressing your comments right now and debugging the integration test(the decommissioned executor doesn't migrate shuffle blocks as expected, so I'm still investigating it.)

Thank you for the further review :)

holdenk commented:
Taking days off is healthy and great! We're not in a rush on this so please take your time. My offer of help with debugging tests was not intended to make you feel like you needed to work on this right away, just I've been in the situation where I'm stuck on a test failure I can't figure out in the past. Let me know when you want me to take another look :)

Ngone51 committed:
fix

Ngone51 committed:
update decommissionExecutors

Ngone51 committed:
update test

Ngone51 committed:
update comment

Ngone51 committed:
update test

Ngone51 committed:
update

Ngone51 committed:
update

SparkQA commented:
**[Test build #129640 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/129640/testReport)** for PR 29817 at commit [`0064ba8`](https://github.com/apache/spark/commit/0064ba862196539726cfb07ab9243ec3c149b871).

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/34244/

SparkQA commented:
Kubernetes integration test status success
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/34244/

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/34244/
Test PASSed.

Ngone51 commented:
cc @holdenk for taking another look. Thanks!

SparkQA commented:
**[Test build #129640 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/129640/testReport)** for PR 29817 at commit [`0064ba8`](https://github.com/apache/spark/commit/0064ba862196539726cfb07ab9243ec3c149b871).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/129640/
Test PASSed.

Ngone51 committed:
remove if

SparkQA commented:
**[Test build #129652 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/129652/testReport)** for PR 29817 at commit [`17bbbd9`](https://github.com/apache/spark/commit/17bbbd9461a898c898b87635e4167ef5e01360d2).

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/34256/

SparkQA commented:
Kubernetes integration test status success
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/34256/

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/34256/
Test PASSed.

SparkQA commented:
**[Test build #129652 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/129652/testReport)** for PR 29817 at commit [`17bbbd9`](https://github.com/apache/spark/commit/17bbbd9461a898c898b87635e4167ef5e01360d2).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/129652/
Test PASSed.

Ngone51 commented:
kindly ping @holdenk

Ngone51 commented:
@holdenk take a look?

HyukjinKwon commented:
Let me merge this in few days if there are no more comments assuming lazy consensus.

Ngone51 commented:
@holdenk ?

dongjoon-hyun commented:
Retest this please.

SparkQA commented:
**[Test build #130119 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/130119/testReport)** for PR 29817 at commit [`17bbbd9`](https://github.com/apache/spark/commit/17bbbd9461a898c898b87635e4167ef5e01360d2).

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/34727/
Test FAILed.

cloud-fan commented on core/src/main/scala/org/apache/spark/deploy/DeployMessage.scala:
```diff
@@ -61,13 +61,34 @@ private[deploy] object DeployMessages {
+   * decommissioning. It's used for the case where decommission is triggered at Worker.
+   *
    * @param id the worker id
-   * @param worker the worker endpoint ref
```
shall we keep the param doc for `workerRef`?

cloud-fan commented on core/src/main/scala/org/apache/spark/deploy/master/Master.scala:
```diff
@@ -245,15 +245,27 @@ private[deploy] class Master(
       } else {
         // We use foreach since get gives us an option and we can skip the failures.
-        idToWorker.get(id).foreach(decommissionWorker)
+        idToWorker.get(id).foreach(w => decommissionWorker(w))
```
unnecessary change?

cloud-fan commented on core/src/main/scala/org/apache/spark/executor/CoarseGrainedExecutorBackend.scala:
```diff
@@ -213,11 +204,31 @@ private[spark] class CoarseGrainedExecutorBackend(
+        driver.foreach { driverRef =>
+          // Tell driver that we are starting decommissioning so it stops trying to schedule us
+          driverNotified = driverRef.askSync[Boolean](ExecutorDecommissioning(executorId))
+          if (driverNotified) decommissionSelf()
```
This part is new. I think it makes sense to skip executor decommissioning if the driver side rejects to do so.

cloud-fan commented on core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedSchedulerBackend.scala:
```diff
@@ -465,72 +464,50 @@ class CoarseGrainedSchedulerBackend(scheduler: TaskSchedulerImpl, val rpcEnv: Rp
+      triggeredByExecutor: Boolean): Seq[String] = withLock {
     // Do not change this code without running the K8s integration suites
-    val executorsToDecommission = executorsAndDecomInfo.filter { case (executorId, decomInfo) =>
-      CoarseGrainedSchedulerBackend.this.synchronized {
```
I have a concern about the scope of the lock. Previously we only hold the lock when we change the mutable status, but now the entire method is locked. Do we have a good reason?

Ngone51 replied:
Because now we not only change the mutable status but also call `scheduler.executorDecommission`. And `scheduler.executorDecommission` requires the lock on `TaskSchedulerImpl`. To avoid potential deadlock, we should use `withLock`.

More context: Previously,  change the mutable status and call `scheduler.executorDecommission` happens in two separate functions. In this PR, we combined these two functions into one function. Therefore, those two operations now are within the same code block and need the `withLock` protection.

SparkQA commented:
**[Test build #130119 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/130119/testReport)** for PR 29817 at commit [`17bbbd9`](https://github.com/apache/spark/commit/17bbbd9461a898c898b87635e4167ef5e01360d2).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/130119/
Test FAILed.

Ngone51 committed:
workerRef doc

Ngone51 committed:
revert unnecessary change

SparkQA commented:
**[Test build #130130 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/130130/testReport)** for PR 29817 at commit [`e625051`](https://github.com/apache/spark/commit/e625051c1ae1d9b7878710aaa4d039b8f577a2c1).

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/34737/

SparkQA commented:
**[Test build #130130 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/130130/testReport)** for PR 29817 at commit [`e625051`](https://github.com/apache/spark/commit/e625051c1ae1d9b7878710aaa4d039b8f577a2c1).
 * This patch **fails due to an unknown error code, -9**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/130130/
Test FAILed.

HyukjinKwon commented:
retest this please

SparkQA commented:
**[Test build #130140 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/130140/testReport)** for PR 29817 at commit [`e625051`](https://github.com/apache/spark/commit/e625051c1ae1d9b7878710aaa4d039b8f577a2c1).

SparkQA commented:
Kubernetes integration test status failure
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/34737/

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/34737/
Test FAILed.

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/34747/

SparkQA commented:
Kubernetes integration test status failure
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/34747/

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/34747/
Test FAILed.

HyukjinKwon commented:
retest this please

SparkQA commented:
**[Test build #130148 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/130148/testReport)** for PR 29817 at commit [`e625051`](https://github.com/apache/spark/commit/e625051c1ae1d9b7878710aaa4d039b8f577a2c1).

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/34755/

SparkQA commented:
**[Test build #130140 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/130140/testReport)** for PR 29817 at commit [`e625051`](https://github.com/apache/spark/commit/e625051c1ae1d9b7878710aaa4d039b8f577a2c1).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/130140/
Test PASSed.

SparkQA commented:
Kubernetes integration test status failure
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/34755/

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/34755/
Test FAILed.

SparkQA commented:
**[Test build #130148 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/130148/testReport)** for PR 29817 at commit [`e625051`](https://github.com/apache/spark/commit/e625051c1ae1d9b7878710aaa4d039b8f577a2c1).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/130148/
Test FAILed.

Ngone51 commented:
retest this please

SparkQA commented:
**[Test build #130159 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/130159/testReport)** for PR 29817 at commit [`e625051`](https://github.com/apache/spark/commit/e625051c1ae1d9b7878710aaa4d039b8f577a2c1).

Ngone51 committed:
update tests

SparkQA commented:
**[Test build #130163 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/130163/testReport)** for PR 29817 at commit [`3c1e033`](https://github.com/apache/spark/commit/3c1e033089e392066425139d7ccb52cd501dcc31).

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/34766/

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/34770/

SparkQA commented:
Kubernetes integration test status failure
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/34766/

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/34766/
Test FAILed.

SparkQA commented:
Kubernetes integration test status success
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/34770/

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/34770/
Test PASSed.

SparkQA commented:
**[Test build #130159 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/130159/testReport)** for PR 29817 at commit [`e625051`](https://github.com/apache/spark/commit/e625051c1ae1d9b7878710aaa4d039b8f577a2c1).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/130159/
Test PASSed.

SparkQA commented:
**[Test build #130163 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/130163/testReport)** for PR 29817 at commit [`3c1e033`](https://github.com/apache/spark/commit/3c1e033089e392066425139d7ccb52cd501dcc31).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/130163/
Test PASSed.

HyukjinKwon commented:
Merged to master.

Ngone51 commented:
thanks all!!

MaxGekk commented on core/src/test/scala/org/apache/spark/storage/BlockManagerDecommissionIntegrationSuite.scala:
```diff
@@ -40,6 +40,46 @@ class BlockManagerDecommissionIntegrationSuite extends SparkFunSuite with LocalS
+        .set(config.DECOMMISSION_ENABLED, true)
+        .set(config.STORAGE_DECOMMISSION_ENABLED, isEnabled)
+      sc = new SparkContext(conf)
+      TestUtils.waitUntilExecutorsUp(sc, 2, 6000)
```
I got test failure in my PR #31131 (the PR is not related to the test I believe):
```
[info] BlockManagerDecommissionIntegrationSuite:
[info] - SPARK-32850: BlockManager decommission should respect the configuration (enabled=false) *** FAILED *** (6 seconds, 165 milliseconds)
[info]   java.util.concurrent.TimeoutException: Can't find 2 executors before 6000 milliseconds elapsed
[info]   at org.apache.spark.TestUtils$.waitUntilExecutorsUp(TestUtils.scala:374)
[info]   at org.apache.spark.storage.BlockManagerDecommissionIntegrationSuite.$anonfun$new$2(BlockManagerDecommissionIntegrationSuite.scala:52)
[info]   at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:23)
```
